src/cdk/observers/private/shared-resize-observer.ts

```diff
@@ -0,0 +1,145 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import {inject, Injectable, NgZone, OnDestroy} from '@angular/core';
+import {Observable, Subject} from 'rxjs';
+import {filter, shareReplay, takeUntil} from 'rxjs/operators';
+import {Platform} from '@angular/cdk/platform';
+
+/**
+ * Handler that logs "ResizeObserver loop limit exceeded" errors.
+ * These errors are not shown in the Chrome console, so we log them to ensure developers are aware.
+ * @param e The error
+ */
+const loopLimitExceededErrorHandler = (e: unknown) => {
+  if (e instanceof Error && e.message === 'ResizeObserver loop limit exceeded') {
+    console.error(
+      `${e.message}. This could indicate a performance issue with your app. See https://github.com/WICG/resize-observer/blob/master/explainer.md#error-handling`,
+    );
+  }
+};
+
+/**
+ * A shared ResizeObserver to be used for a particular box type (content-box, border-box, or
+ * device-pixel-content-box)
+ */
+class SingleBoxSharedResizeObserver {
+  /** Stream that emits when the shared observer is destroyed. */
+  private _destroyed = new Subject<void>();
+  /** Stream of all events from the ResizeObserver. */
+  private _resizeSubject = new Subject<ResizeObserverEntry[]>();
+  /** ResizeObserver used to observe element resize events. */
+  private _resizeObserver?: ResizeObserver;
+  /** A map of elements to streams of their resize events. */
+  private _elementObservables = new Map<Element, Observable<ResizeObserverEntry[]>>();
+
+  constructor(
+    /** The box type to observe for resizes. */
+    private _box: ResizeObserverBoxOptions,
+    /** The platform. */
+    platform: Platform,
+  ) {
+    if (platform.isBrowser) {
+      this._resizeObserver = new ResizeObserver(entries => this._resizeSubject.next(entries));
+    }
+  }
+
+  /**
+   * Gets a stream of resize events for the given element.
+   * @param target The element to observe.
+   * @return The stream of resize events for the element.
+   */
+  observe(target: Element): Observable<ResizeObserverEntry[]> {
+    if (!this._elementObservables.has(target)) {
+      this._elementObservables.set(
+        target,
+        new Observable<ResizeObserverEntry[]>(observer => {
+          const subscription = this._resizeSubject
+            .pipe(filter(entries => entries.some(entry => entry.target === target)))
```

Comment:
Should we move the `filter` down to before the `shareReplay` so we only have a single rxjs operator chain?